### PR TITLE
Closes #2203, #3511: Inline Socorro upload functionality and add supp…

### LINKED ISSUE #2203: MozillaSocorroService: Inline upload functionality
We ran into an issue with the upload caused by incompatible APIs:
https://bugzilla.mozilla.org/show_bug.cgi?id=1530757#c2

`lib-crash` has a `compileOnly` dependency on `GeckoView` to allow us calling `GeckoCrashReporter.sendCrashReport()`. If there's an API change causing incompatibility this just blows up. Since `lib-crash` cannot depend on an actual GeckoView version (it should work with all of them after all) we should bite the bullet and inline the code from `sendCrashReport` and remove the `compileOnly` dependency.

### PR DESCRIPTION
**Pull request by rocketsroger**
…ort for "uncaught exception" crash reports

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 1]
```diff
@@ -20,54 +37,197 @@ typealias GeckoCrashReporter = CrashReporter
...
-        // to upload native code crashes to Socorro if GeckoView is used. If this assumption is going to be wrong in
-        // the future then we can start inlining the functionality here.
-        sendViaGeckoViewCrashReporter(crash)
+        sendReport(null, crash.minidumpPath, crash.extrasPath)
+    }
+
+    @Suppress("ComplexMethod", "LongMethod")
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun sendReport(throwable: Throwable?, miniDumpFilePath: String?, extrasFilePath: String?) {
+        val mURL = URL(defaultServerUrl)
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: The name `mURL` looks wrong here.. :)

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 2]
```diff
@@ -20,54 +37,197 @@ typealias GeckoCrashReporter = CrashReporter
...
 
     override fun report(crash: Crash.NativeCodeCrash) {
-        // GeckoView comes with a crash reporter class that we can use here. For now we are assuming that we only want
-        // to upload native code crashes to Socorro if GeckoView is used. If this assumption is going to be wrong in
-        // the future then we can start inlining the functionality here.
-        sendViaGeckoViewCrashReporter(crash)
+        sendReport(null, crash.minidumpPath, crash.extrasPath)
+    }
+
+    @Suppress("ComplexMethod", "LongMethod")
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: We may be able to split this method into sub methods that add headers, send the meta data, send the crash data..? :)

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 3]
```diff
@@ -20,54 +37,197 @@ typealias GeckoCrashReporter = CrashReporter
...
+
+                var inputLine = it.readLine()
+                while (inputLine != null) {
+                    response.append(inputLine)
+                    inputLine = it.readLine()
+                }
+                it.close()
+                Logger.info("Crash reported to Socorro: $response")
+            }
+        }
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: This part looks like it can throw `IOException`?

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 4]
```diff
@@ -20,54 +37,197 @@ typealias GeckoCrashReporter = CrashReporter
...
+        sendReport(null, crash.minidumpPath, crash.extrasPath)
+    }
+
+    @Suppress("ComplexMethod", "LongMethod")
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun sendReport(throwable: Throwable?, miniDumpFilePath: String?, extrasFilePath: String?) {
+        val mURL = URL(defaultServerUrl)
+        val boundary = generateBoundary()
+        partSet.clear()
+        with(mURL.openConnection() as HttpURLConnection) {
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: The lifetime of some of the objects is hard to follow in this large block. Do we ever close this connection (`disconnect()`)? Are we correctly closing all streams in the error case and are we catching the possible exceptions (`IOException`).

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt [hunk 5]
```diff
@@ -24,27 +25,71 @@ class MozillaSocorroServiceTest {
...
+        assert(service.unescape(test2) == expected2)
+
+        val test3 = "\\\\t"
+        val expected3 = "\t"
+        assert(service.unescape(test3) == expected3)
+
+        val test4 = "\\\\\\\\\\\\t\\\\t\\\\n\\\\\\\\"
+        val expected4 = "\\\t\t\n\\"
+        assert(service.unescape(test4) == expected4)
     }
```
Comment (pocmo) [inline components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt]: I would love to see a test here that spins up a `MockWebServer` and tests/verifies the whole thing (maybe even some error cases?).

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 1 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: oops my bad, will fix.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 2 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Agreed.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 3 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I'll check.  Thanks,

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt (hunk 5 above)
Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt]: OK. I'll look into this.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 4 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Looked into this more.  You're correct. 

> Indicates that other requests to the server are unlikely in the near future. Calling disconnect() should not imply that this HttpURLConnection instance can be reused for other requests.

And since this is a connection that we expect there's no other requests before it times out, we should disconnect.  I'll add it.  Thanks for catching this.

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt (hunk 5 above)
Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt]: Please let me know the two new tests are what you were looking for.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 6]
```diff
@@ -20,54 +39,220 @@ typealias GeckoCrashReporter = CrashReporter
...
+
+            BufferedReader(InputStreamReader(inputStream)).use {
+                val response = StringBuffer()
+                var inputLine = it.readLine()
+                while (inputLine != null) {
+                    response.append(inputLine)
+                    inputLine = it.readLine()
+                }
+
+                it.close()
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: `use {}` will take care of closing for you. :)

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 7]
```diff
@@ -20,54 +39,220 @@ typealias GeckoCrashReporter = CrashReporter
...
+                Logger.info("Crash reported to Socorro: $response")
+            }
+
+            disconnect()
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I wonder if that needs to go into a `finally` block so that we close it in an error case too.

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt (hunk 5 above)
Comment (pocmo) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt]: You can get the request the mockWebServer retrieved to analyze it too. Maybe we can do a quick sanity check that it contains some of the things we expect it to?

Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt]: Sure, I'll try that.  Thanks.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 6 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I'll remove.  Thanks,

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 7 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I see.  Missed this case.  I'll see if I can refactor it better.  Thanks,

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt (hunk 5 above)
Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/service/MozillaSocorroServiceTest.kt]: Got request unzipped into text.  Just using String.contains to find some of the consistent values that we expect to see.  Let me know what you think.  Thanks,

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 8]
```diff
@@ -20,54 +39,215 @@ typealias GeckoCrashReporter = CrashReporter
...
     private val applicationContext: Context,
-    private val appName: String
+    private val appName: String,
+    private val serverUrl: String = defaultServerUrl
 ) : CrashReporterService {
-    override fun report(crash: Crash.UncaughtExceptionCrash) {
-        // Not implemented currently.
+    private val logger = Logger("mozac/MozillaSocorroCrashHelperService")
+    private val startTime = System.currentTimeMillis()
+    val partSet = mutableSetOf<String>()
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: nit: This probably does not need to be visible? Just briefly looking at this I wonder if that needs to be a property at all or whether we could just pass a new mutableSet to every method that needs it from `sendReport()`.

Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Definitely should be private.   I went back and forth on this.  I wasn't sure which way was better, but since you think passing a fresh copy everytime sending a report is better I will implement it that way.  Thanks!

**Comment by rocketsroger** [comment]
bors r=pocmo

**Comment by rocketsroger** [comment]
bors retry

**Comment by rocketsroger** [comment]
bors retry

**Comment by rocketsroger** [comment]
bors r=pocmo

**Comment by Amejia481** [comment]
bors retry

**Comment by rocketsroger** [comment]
bors retry

**Comment by csadilek** [comment]
bors retry

**Comment by pocmo** [comment]
No, you are not.

bors r-

**Comment by pocmo** [comment]
Oh, it's already merged anyways.
